src/memory-analyzer/gdb_api.cpp

```diff
@@ -0,0 +1,469 @@
+
+Module: GDB Machine Interface API
+
+Author: Malte Mues <mail.mues@gmail.com>
+        Daniel Poetzl
+
+\*******************************************************************/
+
+/// \file
+/// Low-level interface to gdb
+
+#ifdef __linux__
+#include <cctype>
+#include <cerrno>
+#include <cstdio>
+#include <cstring>
+#include <regex>
+
+#include "gdb_api.h"
+
+#include <goto-programs/goto_model.h>
+
+#include <util/string_utils.h>
+
+/// Create a gdb_apit object
+///
+/// \param binary the binary to run with gdb
+/// \param log boolean indicating whether gdb input and output should be logged
+gdb_apit::gdb_apit(const char *binary, const bool log)
+  : binary(binary), log(log), gdb_state(gdb_statet::NOT_CREATED)
+{
+}
+
+/// Create a new gdb process for analysing the binary indicated by the member
+/// variable `binary`
+void gdb_apit::create_gdb_process()
+{
+  PRECONDITION(gdb_state == gdb_statet::NOT_CREATED);
+
+  command_log.clear();
+
+  pid_t gdb_process;
+
+  int pipe_input[2];
+  int pipe_output[2];
+
+  if(pipe(pipe_input) == -1)
+  {
+    throw gdb_interaction_exceptiont("could not create pipe for stdin");
+  }
+
+  if(pipe(pipe_output) == -1)
+  {
+    throw gdb_interaction_exceptiont("could not create pipe for stdout");
+  }
+
+  gdb_process = fork();
+
+  if(gdb_process == -1)
+  {
+    throw gdb_interaction_exceptiont("could not create gdb process");
+  }
+
+  if(gdb_process == 0)
+  {
+    // child process
+    close(pipe_input[1]);
+    close(pipe_output[0]);
+
+    dup2(pipe_input[0], STDIN_FILENO);
+    dup2(pipe_output[1], STDOUT_FILENO);
+    dup2(pipe_output[1], STDERR_FILENO);
+
+    dprintf(pipe_output[1], "binary name: %s\n", binary);
+
+    char *arg[] = {const_cast<char *>("gdb"),
+                   const_cast<char *>("--interpreter=mi"),
+                   const_cast<char *>(binary),
+                   NULL};
+
+    dprintf(pipe_output[1], "Loading gdb...\n");
+    execvp("gdb", arg);
+
+    // Only reachable, if execvp failed
+    int errno_value = errno;
+    dprintf(pipe_output[1], "errno in child: %s\n", strerror(errno_value));
+  }
+  else
+  {
+    // parent process
+    gdb_state = gdb_statet::CREATED;
+
+    close(pipe_input[0]);
+    close(pipe_output[1]);
+
+    // get stream for reading the gdb output
+    input_stream = fdopen(pipe_output[0], "r");
+
+    // get stream for writing to gdb
+    output_stream = fdopen(pipe_input[1], "w");
+
+    CHECK_RETURN(most_recent_line_has_tag(R"(~"done)"));
+
+    if(log)
+    {
+      // logs output to `gdb.txt` in the current directory, input is not logged
+      // hence we log it to `command_log`
+      write_to_gdb("-gdb-set logging on");
+      CHECK_RETURN(most_recent_line_has_tag("^done"));
+    }
+
+    write_to_gdb("-gdb-set max-value-size unlimited");
+    CHECK_RETURN(most_recent_line_has_tag("^done"));
+  }
+}
+
+/// Termiante the gdb process and close open streams (for reading from and
+/// writing to gdb)
+void gdb_apit::terminate_gdb_process()
```

Comment:
Do this in a destructor to raise the chances of cleaning up your child on exception?

Reply:
Done.